src/main/java/net/glowstone/entity/GlowLivingEntity.java

```diff
@@ -1109,8 +1107,10 @@ public boolean addPotionEffect(PotionEffect effect, boolean force) {
 
         potionEffects.put(effect.getType(), effect);
 
+        updatePotionEffectsMetadata();
```

Comment:
Should the update calls be done in `pulse` since messages are held until then anyway, instead of in `addPotionEffect` and `removePotionEffect`?

Reply:
I think there is no need for recalculation of potion effect color every pulse

Reply:
There is also no need to recalculate potion effect color every time a potion is added (`addPotionEffects` or multiple `addPotionEffect` calls), before it is made relevant on the network. Perhaps we can queue an update on pulse using a boolean flag, `bPotionEffectsChanged`.